### PR TITLE
Databricks Support

### PR DESCRIPTION
**Pull request by tdimhcsleumas**
This change ports this pr:  https://github.com/flyway/flyway/pull/3681

### CODE REVIEW

File: flyway-database-databricks/src/main/resources/META-INF/services/org.flywaydb.core.extensibility.Plugin [hunk 1]
```diff
@@ -0,0 +1,5 @@
+org.flywaydb.community.database.ignite.thin.IgniteThinDatabaseType
+org.flywaydb.community.database.mysql.tidb.TiDBDatabaseType
+org.flywaydb.community.database.yugabytedb.YugabyteDBDatabaseType
+org.flywaydb.community.database.CommunityDatabaseExtension
```
Comment (Barry-RG) [inline flyway-database-databricks/src/main/resources/META-INF/services/org.flywaydb.core.extensibility.Plugin:4]: These classes are not part of this module and should not be included here.

File: pom.xml [hunk 2]
```diff
@@ -37,8 +37,10 @@
         <module>flyway-database-yugabytedb</module>
         <module>flyway-database-clickhouse</module>
         <module>flyway-database-oceanbase</module>
+        <module>flyway-database-databricks</module>
     </modules>
 
+
```
Comment (Barry-RG) [inline pom.xml:43]: Can you remove this unnecessary new line

**Review by Barry-RG — COMMENTED** [review]
Thank you for porting this change to this repository. 
I have left a couple of review comments to do before merging.

If you have any documentation you could point us to to allow us to smoke test this integration, that would be appreciated.

**Review by Barry-RG — CHANGES_REQUESTED** [review]
My previous comment was meant to be a change request.

File: flyway-database-databricks/src/main/resources/META-INF/services/org.flywaydb.core.extensibility.Plugin (hunk 1 above)
Comment (tdimhcsleumas) [reply flyway-database-databricks/src/main/resources/META-INF/services/org.flywaydb.core.extensibility.Plugin:4]: Will do!

File: flyway-database-databricks/pom.xml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>flyway-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.databricks</groupId>
+            <artifactId>databricks-jdbc</artifactId>
+            <optional>true</optional>
+        </dependency>
```
Comment (Barry-RG) [inline flyway-database-databricks/pom.xml:41]: With the move to this repository this has lost its version. Could you please specify which version of the databricks driver is being used here.

**Review by Barry-RG — CHANGES_REQUESTED** [review]

File: flyway-database-databricks/pom.xml (hunk 3 above)
Comment (Barry-RG) [reply flyway-database-databricks/pom.xml:41]: @tdimhcsleumas is there any update on this?

Comment (tdimhcsleumas) [reply flyway-database-databricks/pom.xml:41]: Hi, in the process of moving at the moment. Will set some time aside to apply updates in the few days.

**Comment by skitscha** [comment]
Can this be merged in now?

**Comment by Barry-RG** [comment]
cc: @tdimhcsleumas @skitscha 
We are currently in the process of trying to test this integration with Databricks and Flyway.

**Comment by Barry-RG** [comment]
cc: @tdimhcsleumas @skitscha 
We are currently in the process of trying to test this integration with Databricks and Flyway. 
If you have any tests or test cases that could help expedite this PR

**Comment by skitscha1** [comment]
I should be able to help test this integration but the build is failing and I cannot contribute to this branch to fix it. @tdimhcsleumas Can you resolve the build issue?

**Comment by tdimhcsleumas** [comment]
@skitscha1 yes. I'll set aside some time in the next few days to fix it.

File: flyway-database-databricks/src/main/java/org/flywaydb/community/database/DatabricksDatabaseExtension.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+public class DatabricksDatabaseExtension implements PluginMetadata {
+    public String getDescription() {
+        return "Community-contributed Databricks database support extension " + readVersion() + " by Redgate";
+    }
+
+    public static String readVersion() {
+        try {
+            return FileUtils.copyToString(
+                    IgniteDatabaseExtension.class.getClassLoader().getResourceAsStream("org/flywaydb/community/database/databricks/version.txt"),
```
Comment (piers-williams) [inline flyway-database-databricks/src/main/java/org/flywaydb/community/database/DatabricksDatabaseExtension.java:33]:
```suggestion
                    DatabricksDatabaseExtension.class.getClassLoader().getResourceAsStream("org/flywaydb/community/database/databricks/version.txt"),
```

File: flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksTable.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * @param jdbcTemplate The JDBC template for communicating with the DB.
+     * @param database     The database-specific support.
+     * @param schema       The schema this table lives in.
+     * @param name         The name of the table.
+     */
+    public DatabricksTable(JdbcTemplate jdbcTemplate, DatabricksDatabase database, DatabricksSchema schema, String name) {
+        super(jdbcTemplate, database, schema, name);
+        this.insertRowLock = new InsertRowLock(jdbcTemplate, 10);
```
Comment (piers-williams) [inline flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksTable.java:22]:
```suggestion
        this.insertRowLock = new InsertRowLock(jdbcTemplate);
```

File: flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabase.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    @Override
+    protected String doGetCurrentUser() throws SQLException {
+        return getMainConnection().getJdbcTemplate().queryForString("SELECT current_user() as user;");
+    }
+
+    @Override
+    public void ensureSupported() {
+        // Always latest Databricks version.
+    }
```
Comment (piers-williams) [inline flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabase.java:31]:
```suggestion
    @Override
    public void ensureSupported(Configuration configuration) {
        // Always latest Databricks version.
    }
```

File: flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabase.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    @Override
+    public boolean supportsDdlTransactions() {
+        // Databricks is non-transactional
+        return false;
+    }
+
+    @Override
+    public boolean supportsChangingCurrentSchema() {
+        return true;
+    }
```
Comment (piers-williams) [inline flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabase.java:42]:
```suggestion
```

File: flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabaseType.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+import org.flywaydb.core.internal.jdbc.StatementInterceptor;
+import org.flywaydb.core.internal.parser.Parser;
+import org.flywaydb.core.internal.parser.ParsingContext;
+import org.flywaydb.core.internal.util.ClassUtils;
+
+import java.sql.Connection;
+import java.sql.Types;
+import java.util.Properties;
+
+public class DatabricksDatabaseType extends BaseDatabaseType {
```
Comment (piers-williams) [inline flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabaseType.java:17]:
```suggestion
public class DatabricksDatabaseType extends BaseDatabaseType implements CommunityDatabaseType {
```

**Review by piers-williams — CHANGES_REQUESTED** [review]
I've taken a look and here are changes that needed to be made to get this building

**Comment by tdimhcsleumas** [comment]
Thanks @piers-williams, I'll implement those changes shortly.

File: flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabaseType.java (hunk 8 above)
Comment (tdimhcsleumas) [reply flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabaseType.java:17]: Not sure about this one given none of the other database types implement `CommunityDatabaseType`.

**Comment by tdimhcsleumas** [comment]
@skitscha @Barry-RG I ran `mvn -B install --file pom.xml` locally and got a successful build.

File: flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabaseType.java (hunk 8 above)
Comment (piers-williams) [reply flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabaseType.java:17]: They all do - it was added 2 months ago. Here is [Ignite](https://github.com/flyway/flyway-community-db-support/blob/6be5b41f1ccbe618b51991dba7f43e2f2f84581c/flyway-database-ignite/src/main/java/org/flywaydb/community/database/ignite/thin/IgniteThinDatabaseType.java#L40)  for reference

Comment (tdimhcsleumas) [reply flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabaseType.java:17]: My bad

Comment (tdimhcsleumas) [reply flyway-database-databricks/src/main/java/org/flywaydb/community/database/databricks/DatabricksDatabaseType.java]: Updated

**Review by piers-williams — APPROVED** [review]

**Review by Barry-RG — APPROVED** [review]

**Pull request merged**